cmake/add_project_smqtk.cmake

```diff
@@ -12,26 +12,60 @@ set( VIAME_PROJECT_LIST ${VIAME_PROJECT_LIST} smqtk )
 if( WIN32 )
   message( FATAL_ERROR "SMQTK not yet supported on WIN32" )
 else()
-  if( VIAME_SYMLINK_PYTHON )
-    set( SMQTK_PIP_CMD
-      pip install --user -e .[postgres] )
-  else()
-    # This is only required for no symlink install without a -e with older
-    # versions of pip, for never versions the above command works with no -e
-    set( SMQTK_PIP_CMD
-      pip install --user file://${VIAME_PACKAGES_DIR}/smqtk\#egg=smqtk[postgres] )
-  endif()
+
+  # Logic for getting site packages is copied from kwiver
+  # (but it uses a new pycmd function that handles indentation -- ooohh. aahh.)
+  pycmd(python_site_packages "
+    from distutils import sysconfig
+    print(sysconfig.get_python_lib(prefix=''))
+  ")
+  message(STATUS "python_site_packages = ${python_site_packages}")
+  get_filename_component(python_sitename ${python_site_packages} NAME)
+
+  pycmd(python_pip_version "import pip; print(pip.__version__)")
+
   set( PYTHON_BASEPATH
     ${VIAME_BUILD_INSTALL_PREFIX}/lib/python${PYTHON_VERSION}${PYTHON_ABIFLAGS} )
-  set( CUSTOM_PYTHONPATH
-    ${PYTHON_BASEPATH}/site-packages:${PYTHON_BASEPATH}/dist-packages:$ENV{PYTHONPATH} )
-  set( CUSTOM_PATH
-    ${VIAME_BUILD_INSTALL_PREFIX}/bin:$ENV{PATH} )
-  set( SMQTK_PYTHON_INSTALL
-    ${CMAKE_COMMAND} -E env PYTHONPATH=${CUSTOM_PYTHONPATH}
-                        env PATH=${CUSTOM_PATH}
-                        env PYTHONUSERBASE=${VIAME_BUILD_INSTALL_PREFIX}
-      ${PYTHON_EXECUTABLE} -m ${SMQTK_PIP_CMD} )
+  set( CUSTOM_SITEPACKAGES ${PYTHON_BASEPATH}/${python_sitename})
+  set( CUSTOM_PYTHONPATH ${CUSTOM_SITEPACKAGES}:$ENV{PYTHONPATH} )
+  set( CUSTOM_PATH ${VIAME_BUILD_INSTALL_PREFIX}/bin:$ENV{PATH} )
+
+  # NOTES: In more recent pip versions, this step is much easier
+  # https://pip.pypa.io/en/stable/news/
+  # Useful commands and the versions they were added in:
+  # version 0.1.4: --install-option
+  # version 1.1.0: --target
+  # version 1.3.0: --root
+  # version 7.0.0: --prefix
+  # version ~8.0.0: fixes issues with no -e and "extras_require"
+
+  if (python_pip_version VERSION_LESS 9.0.0)
+    if( VIAME_SYMLINK_PYTHON )
+      set( SMQTK_PIP_CMD
+        pip install --user -e .[postgres] )
+    else()
+      # This is only required for no symlink install without a -e with older
+      # versions of pip, for never versions the above command works with no -e
+      set( SMQTK_PIP_CMD
+        pip install --user file://${VIAME_PACKAGES_DIR}/smqtk\#egg=smqtk[postgres] )
+    endif()
+    set( SMQTK_PYTHON_INSTALL
+      ${CMAKE_COMMAND} -E env PYTHONPATH=${CUSTOM_PYTHONPATH}
+                          env PATH=${CUSTOM_PATH}
+                          env PYTHONUSERBASE=${VIAME_BUILD_INSTALL_PREFIX}
+        ${PYTHON_EXECUTABLE} -m ${SMQTK_PIP_CMD} )
+  else()
+    # If we are using virtualenvs then a more up to date pip is necessary
+    if( VIAME_SYMLINK_PYTHON )
+      set( SMQTK_PIP_CMD pip install "--prefix=${VIAME_BUILD_INSTALL_PREFIX}" -e .[postgres])
```

Comment:
Instead of setting SMQTK_PIP_CMD independently in both if and else, since this ``if`` is just adding the ``-e``, we can set that string in a variable and unpack it after the statement:
```cmake
...
set( pip_args "" )
if( VIAME_SYMLINK_PYTHON )
    set( pip_args "-e" )
endif()
set( SMQTK_PIP_CMD pip install "--prefix=${VIAME_BUILD_INSTALL_PREFIX}" ${pip_args} .[postgres] )
...
```

Reply:
I would agree if the `-e` was a standalone argument that changed the behavior of pip. However, if you see the pip docs:
```
  -e, --editable <path/url>   Install a project in editable mode (i.e. setuptools "develop mode") from a local project path or a VCS
```
the `-e` is followed by a path, meaning that if futher pip args needed to be specified (e.g. `--no-deps`), then the code would break if that were appended to the end of the args. 

The alternative is to do something like this: 
```cmake
set( pip_args "" )
if( VIAME_SYMLINK_PYTHON )
    set( pip_args "-e .[postgres]" )
else()
    set( pip_args ".[postgres]" )
endif()
set( SMQTK_PIP_CMD pip install "--prefix=${VIAME_BUILD_INSTALL_PREFIX}" ${pip_args} )
```

or this

```cmake
set( pip_path ".[postgres]" )
if( VIAME_SYMLINK_PYTHON )
    set( pip_path "-e ${pip_path}" )
endif()
set( SMQTK_PIP_CMD pip install "--prefix=${VIAME_BUILD_INSTALL_PREFIX}" ${pip_path} )
```
However, I feel like both of these are less clear than the way it exists now.

Reply:
Ahh. I missed the semantic that ``-e`` required a path after it.  What's here is fine then.